Attach to-wireguard to the tc egress of cilium_wg0

jschwinger233 opened the pull request:
This is to handle rev-DNAT when L7 ingress proxy is enabled with wireguard, nodeport, native routing, and KPR.

Consider a pod-to-remote-nodeport connection via L7 ingress proxy, the reply packet will pass: from_lxc@veth -> redirected to L7 proxy -> from_host@cilium_host -> to_host@cilium_net -> to_netdev@eth0 -> redirected to cilium_wg0. This patch ensures the new attached to-wireguard@cilium_wg0 can do the necessary rev-DNAT in this case.

Fixes: https://github.com/cilium/cilium/issues/32899

```release-note
Fixes a missing rev-DNAT issue when wireguard, nodeport, KPR, and L7 proxy are enabled together.
```

jschwinger233 committed:
Attach to-wireguard to the tc egress of cilium_wg0

This is to handle rev-DNAT when L7 ingress proxy is enabled with
wireguard, nodeport, native routing, and KPR.

Consider a pod-to-remote-nodeport connection, the reply packet will
pass: from_lxc@veth -> redirected to L7 proxy -> from_host@cilium_host
-> to_host@cilium_net -> to_netdev@eth0 -> redirected to cilium_wg0.
This patch ensures the new attached to-wireguard@cilium_wg0 can do the
necessary rev-DNAT.

Fixes: https://github.com/cilium/cilium/issues/32899

Signed-off-by: gray <gray.liang@isovalent.com>

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -3063,17 +3063,23 @@ nodeport_rev_dnat_fwd_ipv4(struct __ctx_buff *ctx, bool *snat_done,
+		fib_params.l.family = AF_INET;
+		fib_params.l.ifindex = ctx_get_ifindex(ctx);
+		fib_params.l.ipv4_src = nat_info->address;
+		fib_params.l.ipv4_dst = tuple.daddr;
```
Have you considered adding a new `bpf_wireguard.c` file, so that the program can identify itself as `IS_BPF_WIREGUARD` instead?

jschwinger233 replied:
Yeah definitely, I was just not sure which one is better. Currently busy with debugging broken NS connectivity :fearful:

jschwinger233 replied:
It turns out the skbs are entering tail_handle_snat_fwd_ipv4 tailcall from tail_handle_nat_fwd_ipv4 after failing rev DNAT. For now I added an early return after rev dnat:

```c
static __always_inline int
__handle_nat_fwd_ipv6(struct __ctx_buff *ctx, struct trace_ctx *trace,
		      __s8 *ext_err)
{
	bool snat_done = false;
	int ret;

	ret = nodeport_rev_dnat_fwd_ipv6(ctx, &snat_done, trace, ext_err);
	if (ret != CTX_ACT_OK)
		return ret;

+ #if defined(ENABLE_WIREGUARD)
+ 	if (ctx_get_ifindex(ctx) == WG_IFINDEX) {
+		return CTX_ACT_OK;
+	}
+ #endif /* ENABLE_WIREGUARD */
```

If CI is happy with this, I'll then move the whole thing into a `bpf_wg.c` with specific macros.

julianwiedmann commented on pkg/datapath/loader/base.go:
```diff
@@ -277,6 +277,38 @@ func (l *loader) reinitializeOverlay(ctx context.Context, tunnelConfig tunnel.Co
+	if option.Config.EnableIPv6 {
+		opts = append(opts, "-DENABLE_IPV6")
+	}
+
```
These should all come via `node_config.h`, no?

jschwinger233 replied:
Thank you! No wonder `reinitializeOverlay()` doesn't specify `-DENABLE_IPVX` at all.

dylandreimerink replied:
Yes, you are right, good catch. At least `DENABLE_IPV4` and `DENABLE_IPV6` are. The `opts := []string` part is also present in the reinitializeOverlay so we might need those still.

julianwiedmann replied:
ack, `SECLABEL` and `THIS_INTERFACE_MAC` is needed (and the `CALLS_MAP` ofc).

I'm unsure what the appropriate value for `SECLABEL` would be in WG context. So let's stick with `WORLD_ID` for now :+1:.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -3087,7 +3089,7 @@ nodeport_rev_dnat_fwd_ipv4(struct __ctx_buff *ctx, bool *snat_done,
 		return CTX_ACT_OK;
 
-#if defined(IS_BPF_HOST) && !defined(ENABLE_SKIP_FIB)
+#if defined(IS_BPF_HOST) && !defined(ENABLE_SKIP_FIB) && !defined(IS_BPF_WIREGUARD)
```
Do we really still need to check for both `IS_BPF_HOST` *and* `IS_BPF_WIREGUARD` ?

jschwinger233 replied:
Maybe I don't have to check `IS_BPF_WG` here, instead I can set `ENABLE_SKIP_FIB=1` for bpf_wireguard compiling?

julianwiedmann replied:
`IS_BPF_HOST` is *only* defined in `bpf_host.c`. If we're compiling any other file, this code should be excluded already.

dylandreimerink reviewed (approved):
Loader bits look good to me

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -0,0 +1,48 @@
+#include "lib/nodeport.h"
+#include "lib/clustermesh.h"
+#include "lib/egress_gateway.h"
+
```
I bet we can slim this down quite a bit :grin:

jschwinger233 replied:
Will do

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -0,0 +1,48 @@
+
+	if (validate_ethertype(ctx, &proto))
+		return handle_nat_fwd(ctx, 0, proto, &trace, &ext_err);
+
```
Let's just use `ctx_get_protocol()` here, since we know it's a L3 device?

I think we also need to handle returned errors, and raise a drop notification for them.

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -0,0 +1,48 @@
+		.reason = TRACE_REASON_UNKNOWN,
+		.monitor = 0,
+	};
+
```
Mind adding a `bpf_clear_meta(ctx)` here, for robustness?

jschwinger233 replied:
I added `__u32 src_sec_identity = ctx_load_and_clear_meta(ctx, CB_SRC_LABEL)`. In the meantime, I called `ctx_store_meta(ctx, CB_SRC_LABEL, src->sec_identity)` before bpf_redirect(wg_ifindex) in `bpf/lib/wireguard.h`.

julianwiedmann replied:
hm I don't think that will work / is robust. We have to assume that the kernel stack uses `skb->cb` while processing the packet (before delivering it to the wireguard interface). Hence we typically share such cross-interface context via skb->mark. So I think you would need to use `set_identity_mark()` here.

julianwiedmann reviewed (changes requested):
Thank you! Took a first swing at the BPF parts, only minor things so far.

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -0,0 +1,48 @@
+/* to-wireguard is attached as a tc egress filter to the cilium_wg0 device.
+ */
+__section_entry
+int cil_to_wireguard(struct __ctx_buff *ctx)
```
Maybe we also want to add the new program to `classifyProgramTypes()` ?

jschwinger233 committed:
Add CODEOWNERS for bpf_wireguard.c

Signed-off-by: gray <gray.liang@isovalent.com>

jschwinger233 committed:
bpf/Makefile: Add bpf_wireguard.o for build test

MAX_WIREGUARD_OPTIONS and WIREGUARD_OPTIONS are set in a similar way to
MAX_HOST_OPTIONS and HOST_OPTIONS, with ENABLE_IPSEC replaced by
ENABLE_WIREGUARD.

Signed-off-by: gray <gray.liang@isovalent.com>

julianwiedmann commented on pkg/datapath/loader/base.go:
```diff
@@ -277,6 +277,35 @@ func (l *loader) reinitializeOverlay(ctx context.Context, tunnelConfig tunnel.Co
+		fmt.Sprintf("-DCALLS_MAP=cilium_calls_wireguard_%d", identity.ReservedIdentityWorld),
+
+		"-DENABLE_NODEPORT=1",
+	}
```
This should also come via the [node_config.h](https://github.com/cilium/cilium/blob/99adddceccdc120ad75e6e63c79c9929fa813a59/pkg/datapath/linux/config/config.go#L426), I believe.

bimmlerd reviewed (approved):
CODEOWNERS lgtm

brb reviewed (approved):
Thanks!

jschwinger233 committed:
bpf/complexity-test: Cover bpf_wireguard.c

The files are copied from bpf/complexity-tests/*/bpf_host/*.txt,
followed by s/-DENABLE_IPSEC/-DENABLE_WIREGUARD/.

Signed-off-by: gray <gray.liang@isovalent.com>

julianwiedmann commented on bpf/bpf_wireguard.c:
```diff
@@ -0,0 +1,45 @@
+
+#define ETH_HLEN 0
+#define IS_BPF_WIREGUARD 1
+#define ENABLE_WIREGUARD 1
```
Do we need the `ENABLE_WIREGUARD` ? It should be part of the `node_config.h`, no?

julianwiedmann commented on bpf/complexity-tests/510/bpf_wireguard/1.txt:
```diff
@@ -0,0 +1,37 @@
+-DENABLE_VTEP=1
+-DENABLE_SCTP=1
+-DENABLE_WIREGUARD=1
+-DIS_BPF_WIREGUARD=1
```
Note that IS_BPF_WIREGUARD shouldn't be needed here - it already gets defined by the program itself.

(the other complexity configs don't specify their `IS_BPF_*` value)

julianwiedmann reviewed (approved):
Looks good, thank you!

Just two smaller nits in case you need to rebase once more. Otherwise we can clean those up later.
